File: java/src/main/java/it/giannibombelli/wsc2026/booking/application/commands/PlaceBooking.java
package it.giannibombelli.wsc2026.booking.application.commands;

import it.giannibombelli.wsc2026.common.utils.Require;

import it.giannibombelli.wsc2026.booking.domain.booking.BookingId;
import it.giannibombelli.wsc2026.booking.domain.primitive.GiftCardReference;
import it.giannibombelli.wsc2026.common.application.Command;
import it.giannibombelli.wsc2026.common.domain.primitive.Description;
import it.giannibombelli.wsc2026.common.domain.primitive.Money;


public record PlaceBooking(BookingId aggregateId, Money amount, Description description,
                           GiftCardReference giftCardReference) implements Command<BookingId> {
    public PlaceBooking {
        Require.requireArgument(aggregateId, "bookingId");
        Require.requireArgument(amount, "amount");
        Require.requireArgument(description, "description");
        Require.requireArgument(giftCardReference, "giftCardReference");
    }
}

File: typescript/src/booking/application/commands/placeBooking.ts
// Command per inserire una nuova prenotazione.


import type { Command } from '@/common/application/command.js';
import type { Description } from '@/common/domain/primitive/description.js';
import type { Money } from '@/common/domain/primitive/money.js';
import type { GiftCardReference } from '../../domain/primitive/giftCardReference.js';
import type { BookingId } from '../../domain/booking/bookingId.js';
import { requireArgument } from '@/common/utils/requireArgument.js';

export type PlaceBooking = Command<'PlaceBooking', BookingId> & {
  readonly amount: Money;
  readonly description: Description;
  readonly giftCardReference: GiftCardReference;
};

export function placeBooking(
  aggregateId: BookingId,
  amount: Money,
  description: Description,
  giftCardReference: GiftCardReference
): PlaceBooking {
  requireArgument(aggregateId, 'bookingId');
  requireArgument(amount, 'amount');
  requireArgument(description, 'description');
  requireArgument(giftCardReference, 'giftCardReference');
  return { kind: 'PlaceBooking', aggregateId, amount, description, giftCardReference };
}
